File: android/src/main/java/com/cloudwebrtc/webrtc/utils/AnyThreadSink.java
package com.cloudwebrtc.webrtc.utils;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import io.flutter.plugin.common.EventChannel;

public final class AnyThreadSink implements EventChannel.EventSink {
    private final EventChannel.EventSink eventSink;
    private final Handler handler = new Handler(Looper.getMainLooper());

    public AnyThreadSink(EventChannel.EventSink eventSink) {
        this.eventSink = eventSink;
    }

    @Override
    public void success(Object o) {
        post(()->eventSink.success(o));
    }

    @Override
    public void error(String s, String s1, Object o) {
        post(()->eventSink.error(s, s1, o));
    }

    @Override
    public void endOfStream() {
        post(eventSink::endOfStream);
    }

    private void post(@NonNull Runnable r) {
        if(Looper.getMainLooper() == Looper.myLooper()){
            r.run();
        }else{
            handler.post(r);
        }
    }
}

File: android/src/main/kotlin/com/cloudwebrtc/webrtc/utils/AnyThreadSink.kt
package com.cloudwebrtc.webrtc.utils

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel.EventSink

/**
 * Thread agnostic [EventSink] for sending events from Android side to the Flutter side.
 *
 * @property eventSink Underlying socket, into which all events will be sent.
 */
class AnyThreadSink(private val eventSink: EventSink) : EventSink {
  /** [Runnable] executor on the main Android looper. */
  private val handler = Handler(Looper.getMainLooper())

  override fun success(o: Any) {
    post { eventSink.success(o) }
  }

  override fun error(s: String, s1: String, o: Any) {
    post { eventSink.error(s, s1, o) }
  }

  override fun endOfStream() {
    post { eventSink.endOfStream() }
  }

  /**
   * Schedules the provided [Runnable] on the main Android looper using the [handler].
   *
   * @param r [Runnable] to be scheduled.
   */
  private fun post(r: Runnable) {
    if (Looper.getMainLooper() == Looper.myLooper()) {
      r.run()
    } else {
      handler.post(r)
    }
  }
}
